In [1]:
import pandas as pd

In [2]:
df = pd.read_csv(r'games.csv')

In [3]:
df.head()

id  rated    created_at  last_move_at  turns victory_status winner  \
0  TZJHLljE  False  1.504210e+12  1.504210e+12     13      outoftime  white   
1  l1NXvwaE   True  1.504130e+12  1.504130e+12     16         resign  black   
2  mIICvQHh   True  1.504130e+12  1.504130e+12     61           mate  white   
3  kWKvrqYL   True  1.504110e+12  1.504110e+12     61           mate  white   
4  9tXo1AUZ   True  1.504030e+12  1.504030e+12     95           mate  white   

  increment_code       white_id  white_rating      black_id  black_rating  \
0           15+2       bourgris          1500          a-00          1191   
1           5+10           a-00          1322     skinnerua          1261   
2           5+10         ischia          1496          a-00          1500   
3           20+0  daniamurashov          1439  adivanov2009          1454   
4           30+3      nik221107          1523  adivanov2009          1469   

                                               moves opening_eco  \
0  d4 d5 c4 c6 cxd5 e6 dxe6 fxe6 Nf3 Bb4+ Nc3 Ba5...         D10   
1  d4 Nc6 e4 e5 f4 f6 dxe5 fxe5 fxe5 Nxe5 Qd4 Nc6...         B00   
2  e4 e5 d3 d6 Be3 c6 Be2 b5 Nd2 a5 a4 c5 axb5 Nc...         C20   
3  d4 d5 Nf3 Bf5 Nc3 Nf6 Bf4 Ng4 e3 Nc6 Be2 Qd7 O...         D02   
4  e4 e5 Nf3 d6 d4 Nc6 d5 Nb4 a3 Na6 Nc3 Be7 b4 N...         C41   

                             opening_name  opening_ply  
0        Slav Defense: Exchange Variation            5  
1  Nimzowitsch Defense: Kennedy Variation            4  
2   King's Pawn Game: Leonardis Variation            3  
3  Queen's Pawn Game: Zukertort Variation            3  
4                        Philidor Defense            5

In [4]:
df.tail()

id  rated    created_at  last_move_at  turns victory_status  \
20053  EfqH7VVH   True  1.499791e+12  1.499791e+12     24         resign   
20054  WSJDhbPl   True  1.499698e+12  1.499699e+12     82           mate   
20055  yrAas0Kj   True  1.499698e+12  1.499698e+12     35           mate   
20056  b0v4tRyF   True  1.499696e+12  1.499697e+12    109         resign   
20057  N8G2JHGG   True  1.499643e+12  1.499644e+12     78           mate   

      winner increment_code      white_id  white_rating            black_id  \
20053  white          10+10       belcolt          1691            jamboger   
20054  black           10+0      jamboger          1233  farrukhasomiddinov   
20055  white           10+0      jamboger          1219        schaaksmurf3   
20056  white           10+0  marcodisogno          1360            jamboger   
20057  black           10+0      jamboger          1235               ffbob   

       black_rating                                              moves  \
20053          1220  d4 f5 e3 e6 Nf3 Nf6 Nc3 b6 Be2 Bb7 O-O Be7 Ne5...   
20054          1196  d4 d6 Bf4 e5 Bg3 Nf6 e3 exd4 exd4 d5 c3 Bd6 Bd...   
20055          1286  d4 d5 Bf4 Nc6 e3 Nf6 c3 e6 Nf3 Be7 Bd3 O-O Nbd...   
20056          1227  e4 d6 d4 Nf6 e5 dxe5 dxe5 Qxd1+ Kxd1 Nd5 c4 Nb...   
20057          1339  d4 d5 Bf4 Na6 e3 e6 c3 Nf6 Nf3 Bd7 Nbd2 b5 Bd3...   

      opening_eco                     opening_name  opening_ply  
20053         A80                    Dutch Defense            2  
20054         A41                     Queen's Pawn            2  
20055         D00  Queen's Pawn Game: Mason Attack            3  
20056         B07                     Pirc Defense            4  
20057         D00  Queen's Pawn Game: Mason Attack            3

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20058 entries, 0 to 20057
Data columns (total 16 columns):
id                20058 non-null object
rated             20058 non-null bool
created_at        20058 non-null float64
last_move_at      20058 non-null float64
turns             20058 non-null int64
victory_status    20058 non-null object
winner            20058 non-null object
increment_code    20058 non-null object
white_id          20058 non-null object
white_rating      20058 non-null int64
black_id          20058 non-null object
black_rating      20058 non-null int64
moves             20058 non-null object
opening_eco       20058 non-null object
opening_name      20058 non-null object
opening_ply       20058 non-null int64
dtypes: bool(1), float64(2), int64(4), object(9)
memory usage: 2.3+ MB


In [6]:
df.columns

Index(['id', 'rated', 'created_at', 'last_move_at', 'turns', 'victory_status',
       'winner', 'increment_code', 'white_id', 'white_rating', 'black_id',
       'black_rating', 'moves', 'opening_eco', 'opening_name', 'opening_ply'],
      dtype='object')

In [7]:
df.shape

(20058, 16)

In [12]:
df.isna

<bound method DataFrame.isna of              id  rated    created_at  last_move_at  turns victory_status  \
0      TZJHLljE  False  1.504210e+12  1.504210e+12     13      outoftime   
1      l1NXvwaE   True  1.504130e+12  1.504130e+12     16         resign   
2      mIICvQHh   True  1.504130e+12  1.504130e+12     61           mate   
3      kWKvrqYL   True  1.504110e+12  1.504110e+12     61           mate   
4      9tXo1AUZ   True  1.504030e+12  1.504030e+12     95           mate   
5      MsoDV9wj  False  1.504240e+12  1.504240e+12      5           draw   
6      qwU9rasv   True  1.504230e+12  1.504230e+12     33         resign   
7      RVN0N3VK  False  1.503680e+12  1.503680e+12      9         resign   
8      dwF3DJHO   True  1.503510e+12  1.503510e+12     66         resign   
9      afoMwnLg   True  1.503440e+12  1.503440e+12    119           mate   
10     HgKLWPsz  False  1.503350e+12  1.503350e+12     39           mate   
11     Vf5fKWzI  False  1.503350e+12  1.503350e+12     3

In [13]:
chessdf = df.copy()

In [15]:
#dropping columns with nan values more than 50%
chessdf.dropna(thresh=0.5, inplace=True)

In [16]:
chessdf

id  rated    created_at  last_move_at  turns victory_status  \
0      TZJHLljE  False  1.504210e+12  1.504210e+12     13      outoftime   
1      l1NXvwaE   True  1.504130e+12  1.504130e+12     16         resign   
2      mIICvQHh   True  1.504130e+12  1.504130e+12     61           mate   
3      kWKvrqYL   True  1.504110e+12  1.504110e+12     61           mate   
4      9tXo1AUZ   True  1.504030e+12  1.504030e+12     95           mate   
5      MsoDV9wj  False  1.504240e+12  1.504240e+12      5           draw   
6      qwU9rasv   True  1.504230e+12  1.504230e+12     33         resign   
7      RVN0N3VK  False  1.503680e+12  1.503680e+12      9         resign   
8      dwF3DJHO   True  1.503510e+12  1.503510e+12     66         resign   
9      afoMwnLg   True  1.503440e+12  1.503440e+12    119           mate   
10     HgKLWPsz  False  1.503350e+12  1.503350e+12     39           mate   
11     Vf5fKWzI  False  1.503350e+12  1.503350e+12     38         resign   
12     HRti5mKv  False  1.503350e+12  1.503350e+12     60         resign   
13     2fEjSei6  False  1.503340e+12  1.503340e+12     31         resign   
14     u7i6dOaJ  False  1.504090e+12  1.504100e+12     31           mate   
15     guanvMR5  False  1.504090e+12  1.504090e+12     43         resign   
16     PmpkWkmm  False  1.504090e+12  1.504090e+12     52         resign   
17     EwaK0IsE  False  1.504010e+12  1.504010e+12     66           mate   
18     yrSDozT3  False  1.504010e+12  1.504010e+12    101         resign   
19     x31mXlvc  False  1.503760e+12  1.503760e+12     25         resign   
20     oQklnWWp  False  1.502950e+12  1.502950e+12     14         resign   
21     QFCZwY1f  False  1.502950e+12  1.502950e+12      3         resign   
22     5cgBygpI  False  1.502950e+12  1.502950e+12     17         resign   
23     UhXXBOMY   True  1.502950e+12  1.502950e+12     36         resign   
24     mCij4hBq   True  1.502860e+12  1.502860e+12     13         resign   
25     ifUMWtVj   True  1.502780e+12  1.502780e+12     69           mate   
26     Wf0zuLQQ   True  1.502780e+12  1.502780e+12     43         resign   
27     srz9QfSN   True  1.502780e+12  1.502780e+12     54           mate   
28     NS6ccssb   True  1.502780e+12  1.502780e+12     53         resign   
29     M3vpf2Ki   True  1.502780e+12  1.502780e+12     66           mate   
...         ...    ...           ...           ...    ...            ...   
20028  8HGqt5cf   True  1.504041e+12  1.504042e+12    115           mate   
20029  2PG68cOv   True  1.504013e+12  1.504013e+12     37         resign   
20030  F6akm0V3   True  1.504011e+12  1.504013e+12     82         resign   
20031  Pgee4CKk   True  1.501998e+12  1.502003e+12     50           mate   
20032  MVxE7oz4  False  1.501677e+12  1.501678e+12     53           mate   
20033  Pl7THFmZ  False  1.501677e+12  1.501677e+12     36         resign   
20034  P0FNoKsw   True  1.501438e+12  1.501444e+12    100           draw   
20035  D6pNwmcM   True  1.501014e+12  1.501015e+12     59         resign   
20036  cDSybWIj   True  1.501013e+12  1.501014e+12     43           mate   
20037  iG5MuMz0   True  1.501013e+12  1.501013e+12     32           mate   
20038  xAQi2hl2   True  1.501011e+12  1.501012e+12     36         resign   
20039  NWgLdNx8   True  1.500722e+12  1.500726e+12     57         resign   
20040  yM2DgyaO   True  1.500296e+12  1.500297e+12     87         resign   
20041  B4IEyQyb   True  1.500157e+12  1.500158e+12     58           mate   
20042  5Fx8GBTL   True  1.500156e+12  1.500157e+12     53         resign   
20043  Z490CURF   True  1.500156e+12  1.500156e+12     11         resign   
20044  PVqRI7Ma   True  1.500155e+12  1.500156e+12     37           mate   
20045  e5jFywx3   True  1.500027e+12  1.500031e+12     70         resign   
20046  tM5wivnc   True  1.499871e+12  1.499871e+12     36           mate   
20047  X0DJ1fJu   True  1.499870e+12  1.499871e+12     73           mate   
20048  dnexZDsv   True  1.499869e+12  1.499870e+12     25    

#dropping unrequired columns like id, rated, created_at, last_move_at,increment_code, white_id,black_id, opening_eco,opening_ply


In [20]:
chessdf.drop(columns=['id', 'rated', 'created_at', 'last_move_at','increment_code', 'white_id','black_id', 'opening_eco','opening_ply'], inplace=True)

In [21]:
chessdf.head()

turns victory_status winner  white_rating  black_rating  \
0     13      outoftime  white          1500          1191   
1     16         resign  black          1322          1261   
2     61           mate  white          1496          1500   
3     61           mate  white          1439          1454   
4     95           mate  white          1523          1469   

                                               moves  \
0  d4 d5 c4 c6 cxd5 e6 dxe6 fxe6 Nf3 Bb4+ Nc3 Ba5...   
1  d4 Nc6 e4 e5 f4 f6 dxe5 fxe5 fxe5 Nxe5 Qd4 Nc6...   
2  e4 e5 d3 d6 Be3 c6 Be2 b5 Nd2 a5 a4 c5 axb5 Nc...   
3  d4 d5 Nf3 Bf5 Nc3 Nf6 Bf4 Ng4 e3 Nc6 Be2 Qd7 O...   
4  e4 e5 Nf3 d6 d4 Nc6 d5 Nb4 a3 Na6 Nc3 Be7 b4 N...   

                             opening_name  
0        Slav Defense: Exchange Variation  
1  Nimzowitsch Defense: Kennedy Variation  
2   King's Pawn Game: Leonardis Variation  
3  Queen's Pawn Game: Zukertort Variation  
4                        Philidor Defense

In [37]:
#Count number of wins for whites and blacks 

chessdf.winner.value_counts()

white    10001
black     9107
draw       950
Name: winner, dtype: int64

In [41]:
#Findout number of wins with starting move being e4
chessdf.moves.str.startswith('e4').value_counts('True')

True     0.628079
False    0.371921
Name: moves, dtype: float64

In [51]:
#How many times has the game won by mate by black

res=chessdf[["victory_status","winner"]].groupby(['victory_status', 'winner']).size()
res

victory_status  winner
draw            draw       906
mate            black     2981
                white     3344
outoftime       black      823
                draw        44
                white      813
resign          black     5303
                white     5844
dtype: int64

Conclusion: Game was won by mate by black 2981, drawn 906 times and resigned 5303 times

In [84]:
#what is the least number of turns in which the game was won

res1 = chessdf[["turns","victory_status","winner"]]
res2=res1.sort_values(by='turns')
res2[res2.victory_status=='mate'].head(20)

turns victory_status winner
862        4           mate  black
864        4           mate  black
867        4           mate  black
1632       4           mate  black
847        4           mate  black
14908      4           mate  black
12294      4           mate  black
12296      4           mate  black
12299      4           mate  black
12324      4           mate  black
14303      4           mate  black
14305      4           mate  black
14308      4           mate  black
16032      4           mate  black
7602       4           mate  black
12297      5           mate  white
865        5           mate  white
14306      5           mate  white
3982       6           mate  black
13267      6           mate  black

COnclusion: Black has won by mate with only 4 turns 15 times